src/monitoring_service/cli.py

```diff
@@ -80,24 +113,40 @@ def main(  # pylint: disable=too-many-arguments
     }
     log.info("Contract information", addresses=hex_addresses, start_block=start_block)
 
-    ms = MonitoringService(
-        web3=web3,
-        private_key=private_key,
-        contracts=contracts,
-        sync_start_block=start_block,
-        required_confirmations=confirmations,
-        poll_interval=DEFAULT_POLL_INTERVALL,
-        db_filename=state_db,
-        min_reward=min_reward,
-    )
+    task = None
+    api = None
+    try:
+        service = MonitoringService(
+            web3=web3,
+            private_key=private_key,
+            contracts=contracts,
+            sync_start_block=start_block,
+            required_confirmations=confirmations,
+            poll_interval=DEFAULT_POLL_INTERVALL,
+            db_filename=state_db,
+            min_reward=min_reward,
+        )
+
+        if debug_shell:
+            import IPython
+
+            IPython.embed()
+            return 0
 
-    if debug_shell:
-        import IPython
+        task = gevent.spawn(service.start)
 
-        IPython.embed()
-        return 0
+        log.debug("Starting API")
+        api = MSApi(monitoring_service=service, operator=operator, info_message=info_message)
+        api.run(host=host, port=port)
 
-    ms.start()
+        task.get()
+    finally:
+        log.info("Stopping Monitoring Service...")
+        if api:
+            api.stop()
+        if task:
+            task.kill()
+            task.join()
```

Comment:
Wouldn't `task.get()` be even better here? If it stops due to a GreenletExit  generated by `kill`, it should pass without an error, but raise errors that happened during cleanup.